eclair-core/src/main/scala/fr/acinq/eclair/io/Peer.scala

```diff
@@ -168,54 +168,83 @@ class Peer(val nodeParams: NodeParams, remoteNodeId: PublicKey, wallet: OnChainA
       case Event(SpawnChannelInitiator(c, channelConfig, channelType, localParams, origin), d: ConnectedData) =>
         val channel = spawnChannel(Some(origin))
         c.timeout_opt.map(openTimeout => context.system.scheduler.scheduleOnce(openTimeout.duration, channel, Channel.TickChannelOpenTimeout)(context.dispatcher))
-        val temporaryChannelId = randomBytes32()
-        val channelFeeratePerKw = nodeParams.onChainFeeConf.getCommitmentFeerate(remoteNodeId, channelType, c.fundingSatoshis, None)
-        val fundingTxFeeratePerKw = c.fundingTxFeeratePerKw_opt.getOrElse(nodeParams.onChainFeeConf.feeEstimator.getFeeratePerKw(target = nodeParams.onChainFeeConf.feeTargets.fundingBlockTarget))
-        log.info(s"requesting a new channel with type=$channelType fundingSatoshis=${c.fundingSatoshis}, pushMsat=${c.pushMsat} and fundingFeeratePerByte=${c.fundingTxFeeratePerKw_opt} temporaryChannelId=$temporaryChannelId localParams=$localParams")
-        channel ! INPUT_INIT_FUNDER(temporaryChannelId, c.fundingSatoshis, c.pushMsat, channelFeeratePerKw, fundingTxFeeratePerKw, localParams, d.peerConnection, d.remoteInit, c.channelFlags.getOrElse(nodeParams.channelConf.channelFlags), channelConfig, channelType)
+        val dualFunded = Features.canUseFeature(d.localFeatures, d.remoteFeatures, Features.DualFunding)
+        val temporaryChannelId = if (dualFunded) {
+          Helpers.dualFundedTemporaryChannelId(nodeParams, localParams, channelConfig)
+        } else {
+          randomBytes32()
+        }
+        val fundingTxFeerate = c.fundingTxFeerate_opt.getOrElse(nodeParams.onChainFeeConf.feeEstimator.getFeeratePerKw(target = nodeParams.onChainFeeConf.feeTargets.fundingBlockTarget))
+        val commitTxFeerate = nodeParams.onChainFeeConf.getCommitmentFeerate(remoteNodeId, channelType, c.fundingAmount, None)
+        log.info(s"requesting a new channel with type=$channelType fundingAmount=${c.fundingAmount} dualFunded=$dualFunded pushAmount=${c.pushAmount_opt} fundingFeerate=$fundingTxFeerate temporaryChannelId=$temporaryChannelId localParams=$localParams")
+        channel ! INPUT_INIT_CHANNEL_INITIATOR(temporaryChannelId, c.fundingAmount, dualFunded, commitTxFeerate, fundingTxFeerate, c.pushAmount_opt, localParams, d.peerConnection, d.remoteInit, c.channelFlags_opt.getOrElse(nodeParams.channelConf.channelFlags), channelConfig, channelType)
         stay() using d.copy(channels = d.channels + (TemporaryChannelId(temporaryChannelId) -> channel))
 
-      case Event(msg: protocol.OpenChannel, d: ConnectedData) =>
-        d.channels.get(TemporaryChannelId(msg.temporaryChannelId)) match {
+      case Event(open: protocol.OpenChannel, d: ConnectedData) =>
```

Comment:
Handlers for `protocol.OpenChannel` and `protocol.OpenDualFundedChannel` could be factored.

Reply:
That's what I started with, but then we reverted it after this comment: https://github.com/ACINQ/eclair/pull/2247#discussion_r878212163

There are similarities, but there are enough small differences to make it messy to completely factor together...I'll give it another try though.

Reply:
I have a better middle-ground in https://github.com/ACINQ/eclair/pull/2247/commits/31c3ee911cca907b62b82f5bc8ecedcaa1ad0563, let me know what you think

Reply:
I think it's good enough, we can always revisit later.